pkg/addons/openebs/openebs.go

```diff
@@ -56,63 +57,65 @@ func (o *OpenEBS) HostPreflights() (*v1beta2.HostPreflightSpec, error) {
 	return nil, nil
 }
 
-func (o *OpenEBS) Apply(ctx context.Context) error {
-	loading := pb.Start()
-	loading.Infof("Applying OpenEBS addon")
-	defer loading.Close()
-	version, err := o.latest()
-	if err != nil {
-		return fmt.Errorf("unable to get latest version: %w", err)
-	}
-	if !semver.IsValid(version) {
-		return fmt.Errorf("unable to parse version %s", version)
+func (o *OpenEBS) GetChartFileName() string {
+	return fmt.Sprintf("openebs-%s.tgz", appVersion)
+}
+
+func (o *OpenEBS) GenerateHelmConfig(ctx *cli.Context, isUpgrade bool) ([]dig.Mapping, error) {
+
+	chartConfigs := []dig.Mapping{}
+
+	chartConfig := dig.Mapping{
+		"name":      releaseName,
+		"namespace": o.namespace,
+		"version":   appVersion,
 	}
-	fname := fmt.Sprintf("openebs-%s.tgz", strings.TrimPrefix(version, "v"))
-	hfp, err := charts.FS.Open(fname)
+
+	chartConfig["chartName"] = filepath.Join(defaults.HelmChartSubDir(), o.GetChartFileName())
+
+	valuesStringData, err := yaml.Marshal(helmValues)
 	if err != nil {
-		return fmt.Errorf("unable to find version %s: %w", version, err)
+		return chartConfigs, err
 	}
-	defer hfp.Close()
+	chartConfig["values"] = string(valuesStringData)
+
+	chartConfigs = append(chartConfigs, chartConfig)
 
-	hchart, err := loader.LoadArchive(hfp)
+	err = o.WriteChartFile()
 	if err != nil {
-		return fmt.Errorf("unable to load chart: %w", err)
+		logrus.Fatalf("could not write chart file: %s", err)
 	}
 
-	release, err := o.installedRelease(ctx)
+	return chartConfigs, nil
+}
+
+func (o *OpenEBS) WriteChartFile() error {
+	chartfile := o.GetChartFileName()
+
+	src, err := charts.FS.Open(chartfile)
 	if err != nil {
-		return fmt.Errorf("unable to list openebs releases: %w", err)
+		return fmt.Errorf("could not load chart file: %w", err)
 	}
 
-	if release == nil {
-		o.logger("OpenEBS hasn't been installed yet, installing it.")
-		act := action.NewInstall(o.config)
-		act.Namespace = o.namespace
-		act.ReleaseName = releaseName
-		act.CreateNamespace = true
-		if _, err := act.RunWithContext(ctx, hchart, helmValues); err != nil {
-			return fmt.Errorf("unable to install chart: %w", err)
+	dstpath := filepath.Join(defaults.HelmChartSubDir(), chartfile)
+	dst, err := os.Create(dstpath)
```

Comment:
isn't this going to crash if the file already exists? here the problem with the close on defer also happens (the if err != nil is a bit off thing).